### PR TITLE
🚀 Typescript 4

### PR DESCRIPTION
**Pull request by nicolasgramlich**
# TODO
 - [ ] Fix compile errors
 - [ ] Update utils dependency https://github.com/bluebirds-blue-jay/utils/pull/9

### CODE REVIEW

File: src/utils/clone-with.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { TJSONSchema } from '../types/json-schema';
 import { clone } from './clone';
 
-export function cloneWith<T extends TJSONSchema = TJSONSchema> (schema: T, overrides: Partial<T>): T {
+export function cloneWith<T extends TJSONSchema = TJSONSchema, K extends TJSONSchema = TJSONSchema> (schema: T, overrides: K): T & K {
```
Comment (nicolasgramlich) [inline src/utils/clone-with.ts:4]: I think this is the more correct type, overrides don't need to actually be `Partial<T>`, but instead could be anything, that could possibly overlap with T, but doesn't have to. The result is T & K, which lodash does agree on with me 😅

File: src/utils/merge-with.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import * as Lodash from 'lodash';
 import { TJSONSchema } from '../types/json-schema';
 import { clone } from './clone';
 
-export function mergeWith<T extends TJSONSchema = TJSONSchema> (schema: T, overrides: Partial<T>): T {
+export function mergeWith<T extends TJSONSchema = TJSONSchema, K extends TJSONSchema = TJSONSchema> (schema: T, overrides: K): T & K {
```
Comment (nicolasgramlich) [inline src/utils/merge-with.ts:5]: Same reasoning as above.

File: src/utils/require-properties.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@ import * as Lodash from 'lodash';
 
 export function requireProperties(schema: TObjectJSONSchema, properties: string[], options: { preserveExisting?: boolean } = {}): TObjectJSONSchema {
   if (options.preserveExisting && schema.required) {
-    properties = Lodash.uniq(properties.concat(schema.required));
+    properties = Lodash.uniq(properties.concat(schema.required.map(key => key.toString())));
```
Comment (nicolasgramlich) [inline src/utils/require-properties.ts:7]: Keys can now be symbols and the compiler was happy after adding `toString`. Unsure if this break some edge case 🤔

File: test/utils/merge-with.test.ts [hunk 4]
```diff
@@ -2,20 +2,22 @@ import * as Schema from '../../';
...
+    const barSchema = Schema.object({
       bar: Schema.string()
     }, {
       required: ['bar'],
       additionalProperties: true,
       nullable: true
-    }));
+    });
+
+    const merged = Schema.mergeWith(fooSchema, barSchema);
```
Comment (nicolasgramlich) [inline test/utils/merge-with.test.ts:20]: Just a readability change from when I was debugging the mergeWith changes.

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@bluejay/schema",
-  "version": "1.9.5",
+  "version": "1.10.0",
```
Comment (SylvainEstevez) [inline package.json:3]: That would be 2.0.0 now. Types are including in the package and we're breaking them.

Comment (nicolasgramlich) [reply package.json:3]: Whoops, that must have been from some ancient local changes I had on my computer.

Yeah, definitely 2.0.0 👍

File: src/utils/clone-with.ts (hunk 1 above)
Comment (SylvainEstevez) [reply src/utils/clone-with.ts:4]: `Partial<T>` tells the compiler that the overrides must correspond to the same type as the original schema. You should not be able to add `items: integer()` to a string schema for example (through it will work since schemas are allowed to contain any key/value as per the spec). I do agree that types could be better though, see below.

```typescript
declare function cloneWithOriginal<T extends TJSONSchema>(schema: T, overrides: Partial<T>): T;

const str = string(); // TStringJSONSchema

const clonedOriginal = cloneWithOriginal(str, { minLength: 3 }); // minLength still appears as optional in the cloned type: wrong

declare function cloneWithPullRequest<T extends TJSONSchema, O extends TJSONSchema = TJSONSchema>(schema: T, overrides: O): T & O;

const clonedPullRequest = cloneWithPullRequest(str, { minLength: 3 }); // Here you allow types like TStringJSONScheme & TArrayJSONSchema: wrong

declare function cloneWithBetter<T extends TJSONSchema, O extends Partial<T> = T>(schema: T, overrides: O): T & O;

const clonedBetter = cloneWithBetter(str, { minLength: 3 }); // Here we don't mix types and minLength has become non optional: better
```

File: src/utils/merge-with.ts (hunk 2 above)
Comment (SylvainEstevez) [reply src/utils/merge-with.ts:5]: Same reasoning as above on my side too :)

File: src/utils/require-properties.ts (hunk 3 above)
Comment (SylvainEstevez) [reply src/utils/require-properties.ts:7]: Isn't the problem that the Object type allows for Symbol keys?

```typescript
//  // Seems reasonnable to force string keys on T?
export type TObjectJSONSchema<T extends { [key: string]: any; } = any> = TJSONSchema & { 
  maxProperties?: number;
  minProperties?: number;
  required?: (keyof T)[];
  additionalProperties?: boolean | TJSONSchema;
  definitions?: { [key: string]: TJSONSchema };
  properties?: { [key in keyof T]: TJSONSchema };
  patternProperties?: { [pattern: string]: TJSONSchema };
  dependencies?: { [key: string]: TJSONSchema | string[] };
}
```

File: test/utils/merge-with.test.ts (hunk 4 above)
Comment (SylvainEstevez) [reply test/utils/merge-with.test.ts:20]: 👍

**Review by SylvainEstevez — CHANGES_REQUESTED** [review]

File: src/utils/clone-with.ts (hunk 1 above)
Comment (SylvainEstevez) [reply src/utils/clone-with.ts:4]: You also get autocompletion for the "better" types ;)

Comment (nicolasgramlich) [reply src/utils/clone-with.ts:4]: `O extends Partial<T>` makes more sense. I got confused by the original test case that didn't have attributes in common 🙌 

```
it('should return a cloned object schema', () => {
    const schema = Schema.object({
      foo: Schema.string()
    }, {
      required: ['foo'],
      additionalProperties: true
    });

    const cloned = Schema.cloneWith(schema, Schema.object({ bar: Schema.string() }, {
      additionalProperties: true,
      nullable: true
    }));

    expect(cloned).to.deep.equal({
      type: ['object', 'null'],
      required: ['foo'],
      additionalProperties: true,
      properties: {
        bar: { type: 'string' }
      }
    });
  });
```

**Comment by nicolasgramlich** [comment]
@SylvainEstevez so that cloneWith/mergeWith change actually doesn't actually work like that because (or well the test cases might need adjustment):

Screen Shot 2020-09-14 at 4 57 23 PM
Screen Shot 2020-09-14 at 4 57 16 PM

Basically it prevents one from "merging in" / "cloning in" new properties that aren't on the source object already, which I think isn't desired 🤔

File: src/utils/require-properties.ts (hunk 3 above)
Comment (nicolasgramlich) [reply src/utils/require-properties.ts:7]: @SylvainEstevez btw, what about `unknown` in this case:

```Typescript
export type TObjectJSONSchema<T extends { [key: string]: unknown } = any> = TJSONSchema & {
```

Might not make a difference here in this case, but feels like the right choice.

**Comment by nicolasgramlich** [comment]
Latest build should have exactly these fail on travis: https://travis-ci.org/github/bluebirds-blue-jay/schema/builds/727219781

File: src/utils/require-properties.ts (hunk 3 above)
Comment (SylvainEstevez) [reply src/utils/require-properties.ts:7]: `unknown` is probably a better choice for libraries indeed 👍

**Review by SylvainEstevez — APPROVED** [review]
